### PR TITLE
[DSV4] Chunk the indexer MQA logits by query rows under a free-memory budget

### PR DESCRIPTION
**Pull request by shikicloud**
## Motivation

The DeepSeek-V4 C4 indexer scores every query row of a prefill chunk against every compressed (c4) position of the context into one fp32 logits matrix, `[query_rows, align256(max_compressed_seq_len)]`, allocated inside DeepGEMM (`fp8_paged_mqa_logits` on the paged path, `fp8_mqa_logits` on the nonpaged path) once per layer. Nothing sized by `--mem-fraction-static` accounts for it, and it grows linearly with context:

| shape | logits |
|---|---|
| 4096 rows x 372K ctx (issue repro, H100 TP4) | 1.42 GiB |
| 4096 rows x 800K ctx (last chunk of the same request) | 3.05 GiB |
| 8192 rows x 1M ctx (default chunked-prefill, nonpaged path) | ~8 GiB |

So a long prefill deterministically OOMs mid-request on every TP rank once the KV pool has consumed the headroom. The DSA (V3.2) indexer already solves the same problem in `_get_topk_ragged` by chunking the query rows under a free-memory budget; the DSV4 indexer had no equivalent on either path, while SM120 already chunked the paged path (fixed 4096 rows, DeepGEMM metadata kernel cap) and ROCm FP4 chunked into a pooled buffer.

## Modifications

- `dsa/utils.py`: DSA budget logic (`_get_mqa_logits_budget_bytes`, `_should_chunk_mqa_logits`) extracted into shared helpers; `dsa_indexer.py` delegates, behavior unchanged. Knob stays `SGLANG_DSA_MQA_LOGITS_FREE_MEM_FRACTION`; no new env var.
- `dsv4/metadata.py`: `PagedIndexerMetadata.__post_init__` decides once per forward: one `mem_get_info`, `rows_per_chunk = min(SM120 cap, rows the budget allows)`, then a DeepGEMM schedule and a top-k v2 plan per chunk. The SM120 list mechanism becomes the generic one. No chunking under prefill CUDA graph, stream capture, BCG/piecewise, or below 8M elements.
- `dsv4/indexer.py`: the SM120 list loop steps by `rows_per_chunk` and uses per-chunk top-k plans (slices previously reused the full-batch plan on CUDA); the nonpaged path gathers K once and chunks `fp8_mqa_logits` by rows.

## Accuracy Tests

- CPU unit tests: 21 cases pass (`test_dsv4_indexer_chunk.py`, `test_dsv4_nonpaged_indexer.py`, existing `test_dsa_mqa_logits_chunking.py`).
- GB200 kernel check at the issue shape (4096 x 93000 c4, cluster-path rows included): chunked vs unchunked `fp8_paged_mqa_logits` / `fp8_mqa_logits` logits bit-identical; per-chunk top-k v2 plans select the same pages as the full plan.
- E2E, DeepSeek-V4-Flash-0731 TP4 GB200: forced chunking (fraction 0.0005) vs default, prompt logprob deviation inside the run-to-run noise floor of two default launches (mean |dlogprob| ~0.08, top-1 agreement ~0.9), on both the paged (cps 4096) and nonpaged (cps 16384) paths; chunk logs 24/16 vs 0.
- Issue shape: 692K-token prompt, `--mem-fraction-static 0.95` (3.7 GB free), prefill completes in 49 s with no OOM (late steps chunk to 411 rows x 173K cols).

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [x] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [x] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [x] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [x] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review and Merge Process

1. Ping Merge Oncalls to start the process. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - Common commands include `/tag-and-rerun-ci`, `/tag-run-ci-label`, `/rerun-failed-ci`
4. After green CI and required approvals, ask Merge Oncalls or people with Write permission to merge the PR.

---
### CI States

Latest PR Test (Base): :white_check_mark: [Run #35405295077](https://github.com/sgl-project/sglang/actions/runs/35405295077)
Latest PR Test (Extra): :white_check_mark: [Run #35405873102](https://github.com/sgl-project/sglang/actions/runs/35405873102)
Latest PR Test (AMD ROCm 10): :x: [Run #35405295029](https://github.com/sgl-project/sglang/actions/runs/35405295029)

### CODE REVIEW

File: python/sglang/srt/layers/attention/dsa/utils.py [hunk 1]
```diff
@@ -250,3 +259,64 @@ def fp8_mqa_logits_make_fused_kv(
             kv_scales[blk].float().contiguous().view(torch.uint8).reshape(-1)
         )
     return fused.view(num_phys_blocks, block_kv, 1, per_token_size)
+
+
+MQA_LOGITS_BYTES_PER_ELEM = 4
```
Comment (Fridge003) [inline python/sglang/srt/layers/attention/dsa/utils.py]: Can we put these utils in a standalone file (maybe called  `python/sglang/srt/layers/attention/mqa_logits_utils.py`), since they can be shared across dsa/dsv4/dsv41

File: python/sglang/srt/layers/attention/dsa/dsa_indexer.py
Comment (Fridge003) [inline python/sglang/srt/layers/attention/dsa/dsa_indexer.py]: Can we also extract `_should_chunk_mqa_logits` to a common space

File: test/registered/unit/layers/test_dsv4_indexer_chunk.py [hunk 2]
```diff
@@ -0,0 +1,276 @@
+"""Unit tests for the DSV4 indexer's per-forward MQA-logits row chunking."""
```
Comment (Fridge003) [inline test/registered/unit/layers/test_dsv4_indexer_chunk.py]: No need to create a new unit test. Please put tests related to dsv4 chunk indexer in test/registered/unit/layers/test_dsv4_nonpaged_indexer.py

File: python/sglang/srt/layers/attention/dsa/utils.py (hunk 1 above)
Comment (shikicloud) [reply python/sglang/srt/layers/attention/dsa/utils.py]: Done. The constants and helpers now live in python/sglang/srt/layers/attention/mqa_logits_utils.py.

File: python/sglang/srt/layers/attention/dsa/dsa_indexer.py
Comment (shikicloud) [reply python/sglang/srt/layers/attention/dsa/dsa_indexer.py]: Done. I`ndexer._should_chunk_mqa_logits` is removed. The whole decision is now in `mqa_logits_utils.py`.

File: test/registered/unit/layers/test_dsv4_indexer_chunk.py (hunk 2 above)
Comment (shikicloud) [reply test/registered/unit/layers/test_dsv4_indexer_chunk.py]: Done. The chunking tests moved into `test_dsv4_nonpaged_indexer.py`.

**Review by Fridge003 — APPROVED** [review]

**Comment by Fridge003** [comment]
Extra tests passed https://github.com/sgl-project/sglang/actions/runs/35192680570

**Review by kpham-sgl — APPROVED** [review]
